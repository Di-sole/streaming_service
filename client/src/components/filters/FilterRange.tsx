import Image from 'next/image';
import styles from './FilterRange.module.scss';
import { useEffect, useState } from 'react';
import greaterThan from '../../../public/icons/greater-than.svg';
import { useTranslation } from 'next-i18next';
import { useAppDispatch, useAppSelector } from '../../store/hooks/redux';
import { checkFiltersStatus } from '../../utils/functions';
import axios from 'axios';
import { addFilteredMovies, ratingFilterAdded, votesFilterAdded } from '../../store/slices/moviesSlice';
import IMovies from '@/models/IMovies';
import IGenre from '@/models/IGenre';
import ICountry from '@/models/ICountry';
interface IRating {
    category: 'ratingKp' | 'votesKp';
    image: string;
    limit: number;
    step: number;
}

interface IFiltered {
    film: IMovies;
    genres: IGenre[];
    countries: ICountry[];
}

const FilterRange = ({ category, image, limit, step }: IRating) => {
    const { t } = useTranslation('moviesPage');
    const dispatch = useAppDispatch();
    const [ratingValue, setRatingValue] = useState(0);
    const allFilters = useAppSelector((state) => state.movies.filters);
    const joinedQuery = `votesKp=${allFilters.votesKp}&ratingKp=${allFilters.ratingKp}`;

    useEffect(() => {
        category === 'ratingKp' ? dispatch(ratingFilterAdded(ratingValue)) : dispatch(votesFilterAdded(ratingValue));
    }, [ratingValue]);

    useEffect(() => {
        const fetchRating = async () => {
            try {
                const response = await axios.get(`http://localhost:6125/movies?${joinedQuery}&limit=1000`);
                dispatch(addFilteredMovies(response.data.docs[0].page));
            } catch (err) {
                console.log(err);
            }
        };
        ratingValue !== 0 && fetchRating();
    }, [ratingValue]);

    useEffect(() => {
        if (checkFiltersStatus(allFilters)) {
            setRatingValue(0);
        }
    }, [allFilters]);
    return (
        <div className={styles.container}>
            <div className={styles.container__ratingContent}>
                <div
                    className={[
                        styles.container__ratingValue,
                        ratingValue === 0 || ratingValue === limit ? styles.container__ratingValue_notSelected : null,
                    ].join(' ')}
                >
                    <Image src={image} height={26} width={26} alt="Рейтинг" />
                    <Image src={greaterThan} height={18} width={18} alt="Знак больше" />
                    <span>{ratingValue === 0 ? t('filterPanel.notSelected') : ratingValue.toLocaleString()}</span>
                </div>
                <div>
                    <input
                        type="range"
                        max={limit}
                        value={ratingValue}
                        step={step}
                        onChange={(e) => setRatingValue(e.target.valueAsNumber)}
                    />
                </div>
            </div>
        </div>
    );
};

export default FilterRange;
