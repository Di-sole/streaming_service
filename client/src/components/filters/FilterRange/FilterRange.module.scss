.container {
    display: flex;
    justify-content: center;

    &__ratingContent {
        display: flex;
        flex-direction: column;
        padding-top: 12px;
    }

    &__ratingValue {
        display: flex;
        justify-content: center;
        align-items: center;

        margin-bottom: 15px;

        img {
            margin-right: 5px;
        }

        span {
            font-size: 18px;
            font-weight: 500;
            color: #fff;

            text-align: center;
        }

        &_notSelected > img:nth-child(2) {
            display: none;
        }
    }

    input[type='range'] {
        -webkit-appearance: none;
        appearance: none;
        width: 100%;
    }

    input[type='range']::-webkit-slider-runnable-track {
        width: 300px;
        height: 4px;
        background: #ddd;
        border: none;
        border-radius: 15px;
    }

    input[type='range']::-webkit-slider-thumb {
        -webkit-appearance: none;
        border: none;
        height: 12px;
        width: 12px;
        border-radius: 50%;
        background: #fff;
        margin-top: -4px;
    }

    input[type='range']:focus {
        outline: none;
    }

    input[type='range']:focus::-webkit-slider-runnable-track {
        background: #ccc;
    }
}
