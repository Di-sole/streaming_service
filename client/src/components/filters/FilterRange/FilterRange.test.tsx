import { render, screen } from '@testing-library/react';
import FilterRange from './FilterRange';
import * as reduxHooks from '../../../store/hooks/redux';
import ratingIcon from '@/../public/icons/rating.svg';
import votesIcon from '../../../public/icons/userRank.svg';

jest.mock('../../../store/hooks/redux');
const mockedSelector = jest.spyOn(reduxHooks, 'useAppSelector');
const useDispatch = jest.spyOn(reduxHooks, 'useAppDispatch');
const mockDispatch = jest.fn();
const filters = {
    genres: [],
    countries: [],
    ratingKp: 0,
    votesKp: 0,
    director: '',
    actor: '',
};

jest.mock('next-i18next', () => ({
    useTranslation: () => {
        return {
            t: (str: string) => str,
            i18n: {
                changeLanguage: () => new Promise(() => {}),
            },
        };
    },
}));

describe('FILTERRANGE SNAPSHOTS TESTS', () => {
    test('"category" ratingKp FilterRange should not be changed', () => {
        mockedSelector.mockReturnValue(filters);
        useDispatch.mockReturnValue(mockDispatch);
        render(<FilterRange category="ratingKp" image={ratingIcon} limit={5} step={1} />);

        const div = screen.getByTestId('filterRange');
        expect(div).toMatchSnapshot();
    });

    test('"category" votesKp FilterRange should not be changed', () => {
        mockedSelector.mockReturnValue(filters);
        useDispatch.mockReturnValue(mockDispatch);
        render(<FilterRange category="votesKp" image={votesIcon} limit={5} step={1} />);

        const div = screen.getByTestId('filterRange');
        expect(div).toMatchSnapshot();
    });
});
