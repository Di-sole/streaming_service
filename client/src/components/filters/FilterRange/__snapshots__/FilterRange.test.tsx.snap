// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FILTERRANGE SNAPSHOTS TESTS "category" ratingKp FilterRange should not be changed 1`] = `
<div
  class="container"
  data-testid="filterRange"
>
  <div
    class="container__ratingContent"
  >
    <div
      class="container__ratingValue container__ratingValue_notSelected"
    >
      <img
        alt="Рейтинг"
        data-nimg="1"
        decoding="async"
        height="26"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
        style="color: transparent;"
        width="26"
      />
      <img
        alt="Знак больше"
        data-nimg="1"
        decoding="async"
        height="18"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
        style="color: transparent;"
        width="18"
      />
      <span>
        filterPanel.notSelected
      </span>
    </div>
    <div>
      <input
        max="5"
        step="1"
        type="range"
        value="0"
      />
    </div>
  </div>
</div>
`;

exports[`FILTERRANGE SNAPSHOTS TESTS "category" votesKp FilterRange should not be changed 1`] = `
<div
  class="container"
  data-testid="filterRange"
>
  <div
    class="container__ratingContent"
  >
    <div
      class="container__ratingValue container__ratingValue_notSelected"
    >
      <img
        alt="Рейтинг"
        data-nimg="1"
        decoding="async"
        height="26"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
        style="color: transparent;"
        width="26"
      />
      <img
        alt="Знак больше"
        data-nimg="1"
        decoding="async"
        height="18"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
        style="color: transparent;"
        width="18"
      />
      <span>
        filterPanel.notSelected
      </span>
    </div>
    <div>
      <input
        max="5"
        step="1"
        type="range"
        value="0"
      />
    </div>
  </div>
</div>
`;
